### ISSUE TITLE
e2e test failure: Kubectl client Simple pod should support exec through an HTTP proxy

### ISSUE THREAD

**Issue by erictune** [kubernetes/kubernetes#15713]
@ncdc @ashcrow 

#11694 was merged a few hours ago

It added the following test:

"Kubectl client Simple pod should support exec through an HTTP proxy"

I saw the following flake in several of our e2e runs:

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:359
Oct 15 08:56:39.743: Unable to create static kubectl. Error=exit status 2, Output=
```

Any idea if this might be a flaky test?

**Comment by ncdc** [comment]
@erictune I'll defer to @ashcrow on this. Is there any more info about the failure?

**Comment by erictune** [comment]
No, e2e doesn't save much output.  

Suggest running just that test in a loop to see if it is flaky.

**Comment by ashcrow** [comment]
It shouldn't be flaky. I've run it many times in a gce cluster as well as local. The only way I could think that it may flake would be if the jenkins e2e builders are not all create the same and some are missing golang and or make.

However, I'll take a look and see if maybe there is some possible case where an error on success occurs if the same call is made two times.

**Comment by ashcrow** [comment]
```
$ make -C $WORKSPACE/kubernetes/examples/kubectl-container
make: Entering directory '/home/steve/Tech/GITHUB/kubernetes/examples/kubectl-container'
KUBE_STATIC_OVERRIDES="kubectl" ../../hack/build-go.sh cmd/kubectl; cp ../../_output/local/bin/linux/amd64/kubectl .
+++ [1015 13:28:46] Building go targets for linux/amd64:
    cmd/kubectl
We can rule out building when an already built static binary has been created.

+++ [1015 13:28:50] Placing binaries
make: Leaving directory '/home/steve/Tech/GITHUB/kubernetes/examples/kubectl-container'
$ echo $?
0
$ make -C $WORKSPACE/kubernetes/examples/kubectl-container
make: Entering directory '/home/steve/Tech/GITHUB/kubernetes/examples/kubectl-container'
make: 'kubectl' is up to date.
make: Leaving directory '/home/steve/Tech/GITHUB/kubernetes/examples/kubectl-container'
$ echo $?
0
```

**Comment by ncdc** [comment]
Does `exit status 2` mean `ENOENT` in this case?

**Comment by ashcrow** [comment]
@ncdc good question. I'll see if I get the error without the path.

**Comment by jlowdermilk** [comment]
I ran the make command in the jenkins job workspace, and it looks like the problem is that `make -C /examples/kubectl-container/` doesn't work when run from an extracted kubernetes.tar release bundle. It's trying to find `hack/build-go.sh` which isn't included, and looking for the kubectl binary in `_output/local/` instead of `platforms/`. It can either skip or copy the already built binary if the test is running from a built release directory.

**Comment by ncdc** [comment]
@jlowdermilk it needs to be a static binary b/c we're running kubelet in a container as part of this test.

I'm curious how this passed the Jenkins GCE e2e tests before?

**Comment by ashcrow** [comment]
@jlowdermilk hrm. Thank you for looking! I wonder how this passed during the review. Unfortunately, we can't use the prebuilt kubectl since it is not static.

**Comment by ncdc** [comment]
@jlowdermilk what is the best way to fix this?

**Comment by jlowdermilk** [comment]
The reason it passed review is because the pr builder actually pulls the repo branch and builds it, so those files are present. I think fixing may require bundling static kubectl in the release.

**Comment by ncdc** [comment]
Ok, so have both a dynamic and a static build of kubectl in the release?

**Comment by ashcrow** [comment]
That could work. We could make every build produce the current dyanmic ```kubectl``` as well as a ```kubectl.static```.

**Comment by jlowdermilk** [comment]
Why ship only the static kubectl? I thought that was the plan when making it static.

**Comment by ashcrow** [comment]
@jlowdermilk you mean why not ship just a static ```kubectl```? It was proposed but not accepted in #11694.

**Comment by ashcrow** [comment]
Just so people are aware, I'm going through the bash build stuff now to try to shoehorn in a static build along side the dynamic build.

**Comment by ashcrow** [comment]
@jlowdermilk I noticed that when making the release tars the ```kubernetes.tar.gz``` has ```platforms/linux/386/kubectl``` which is already built as static. Would this path be available when pulling the repo branch doesn't occur?

**Comment by jlowdermilk** [comment]
Yep, both get exploded in the test workspace, and tests run from the `kubernetes` dir. So if the http exec test can use `platforms/linux/386/kubectl`, then no need to add another build target.

**Comment by ashcrow** [comment]
#15787 has been merged which should fix this issue.

**Comment by fabioy** [comment]
#15787 caused gce to go yellow, with this failure:

Kubectl client Simple pod should support exec through an HTTP proxy

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:378
Oct 17 16:05:31.702: Can't build static kubectl due to missing hack/build-go.sh. Error=stat /var/lib/jenkins/jobs/kubernetes-e2e-gce/workspace/hack/build-go.sh: no such file or directory

 Reverted in #15821.

**Comment by ashcrow** [comment]
@fabioy That case should only occur if there is no 386 ```kubectl``` in the release tars AND what is required to build a static ```kubectl``` does not exist. Can you give some information on the GCE setup you had yellow show up?

**Comment by ashcrow** [comment]
@fabioy @jlowdermilk is there possibly a third type of jenkins e2e run that doesn't include the release tars nor a full checkout?

**Comment by fabioy** [comment]
I'm not sure (I'm not as familiar with the test setups). There's a separate issue filed to follow up with this (i.e. why merge-bot accepted it, but broke e2e): #15848.

**Comment by ashcrow** [comment]
PR #15875 has been merged and should fix this issue.

**Comment by jlowdermilk** [comment]
Shiny and [green](http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/lastCompletedBuild/testReport/(root)/Kubernetes%20e2e%20suite/Kubectl_client_Simple_pod_should_support_exec_through_an_HTTP_proxy/history/). Fixed by #15875
